### PR TITLE
Change batch_flatten to compute shape at construction time

### PR DESCRIPTION
**Pull request by t-vi**
Hello,

not sure if this is better as a PR or an issue:
Currently, batch_flatten and the Flatten layer drop backend shape information, at least for Tensorflow.
I ran into trouble with that a couple of times, notably when using Lambda layers for experiments or when trying to do an auxiliary computation in tensorflow.

This changes batch_flatten to compute the shape passed to reshape
at construction time (via K.int_shape) rather than using a
symbolic shape (via K.shape).
This has the advantage that the output tensor of a Flatten layer
has a (native) shape.

This can be important when using the output
tensor directly, e.g. with a Lambda layer after the Flatten
layer or when using output in the backend directly.
In particular, this change would solve the issue of this SO
question regarding a Flatten layer followed a Lambda layer:
http://stackoverflow.com/questions/37035944/unable-to-connect-lambda-layer-in-keras-directly-to-flatten-layer-without-an-exc

As far as I can see, the tests all pass. I have not found users of batch_flatten other than the Flatten layer in Keras.

Thank you for your consideration.

Best regards

Thomas

### CODE REVIEW

**Comment by fchollet** [comment]
This change requires input shapes to be fully defined (use of `int_shape`). It would raise an exception with any variable-size dimension.
